### PR TITLE
h3 config examples

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: n/a
Testing: examples config test, manual testing
Docs Changes: n/a
Release Notes: n/a
Part of #12923
Co-authored-by: Michael Payne <m@m17e.org>

### CODE REVIEW

**Comment by alyssawilk** [comment]
Hey Dan,

one weird thing, if I run bazel-bin/source/exe/envoy-static --config-path configs/envoyproxy_io_proxy_http3_downstream.template.yaml *--concurrency 1*
It works (/git/curl$ ./src/curl  -vv --http3 https://127.0.0.1:10000 gets results)
If I run bazel-bin/source/exe/envoy-static --config-path configs/envoyproxy_io_proxy_http3_downstream.template.yaml 
I get consistent failures : https://pastebin.com/eU4DeqPd

Is this due to something wrong with my example config, or like @moderation posited is there something going wrong with QUIC with concurrency > 1 ?
Basically if you can't tell me what I did wrong I'll file a tracking issue and you can repro once this lands :-)

**Comment by alyssawilk** [comment]
Actually I'm going to file a tracking issue either way - either we have a bug with concurrency or we have a bug with config validation (given it lets me run but doesn't work)

**Comment by moderation** [comment]
Cross reference - https://github.com/envoyproxy/envoy/issues/15845

File: configs/envoyproxy_io_proxy_http3_downstream.template.yaml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+          protocol: UDP
+          address: 0.0.0.0
+          port_value: 10000
+      reuse_port: true
+      udp_listener_config:
+        quic_options: {}
+        downstream_socket_config:
+          prefer_gro: true
+      filter_chains:
+        transport_socket:
```
Comment (yuval-k) [inline configs/envoyproxy_io_proxy_http3_downstream.template.yaml:21]: i think a hyphen is missing here? i.e.
`- transport_socket:`

Comment (alyssawilk) [reply configs/envoyproxy_io_proxy_http3_downstream.template.yaml:21]: the example configs test passed (locally for me - CI failed catastrophically for unrelated reasons) so I assume it works either way?

**Comment by alyssawilk** [comment]
Ah, I do have to fix config examples to not choke on H3 configs though.
/wait

File: configs/google_com_http3_upstream_proxy.yaml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    socket_address:
+      protocol: TCP
+      address: 0.0.0.0
+      port_value: 9901
+static_resources:
+  listeners:
+  - name: listener_0
+    address:
+      socket_address:
+        protocol: TCP
```
Comment (danzh2010) [inline configs/google_com_http3_upstream_proxy.yaml:12]: UDP?

File: configs/envoyproxy_io_proxy_http3_downstream.template.yaml (hunk 1 above)
Comment (moderation) [reply configs/envoyproxy_io_proxy_http3_downstream.template.yaml:21]: It works with or without a hyphen (which is a bit surprising to me) ¯\_(ツ)_/¯

File: configs/google_com_http3_upstream_proxy.yaml (hunk 2 above)
Comment (moderation) [reply configs/google_com_http3_upstream_proxy.yaml:12]: I think the idea here is a standard TCP listener that makes a h3 upstream call to h3 enabled www.google.com

**Comment by danzh2010** [comment]
> Hey Dan,
> 
> one weird thing, if I run bazel-bin/source/exe/envoy-static --config-path configs/envoyproxy_io_proxy_http3_downstream.template.yaml _--concurrency 1_
> It works (/git/curl$ ./src/curl  -vv --http3 https://127.0.0.1:10000 gets results)
> If I run bazel-bin/source/exe/envoy-static --config-path configs/envoyproxy_io_proxy_http3_downstream.template.yaml
> I get consistent failures : https://pastebin.com/eU4DeqPd
> 
> Is this due to something wrong with my example config, or like @moderation posited is there something going wrong with QUIC with concurrency > 1 ?
> Basically if you can't tell me what I did wrong I'll file a tracking issue and you can repro once this lands :-)

It not obvious to me what's going wrong here. The config looks reasonable. I'll try it out locally.

File: configs/google_com_http3_upstream_proxy.yaml (hunk 2 above)
Comment (danzh2010) [reply configs/google_com_http3_upstream_proxy.yaml:12]: Oh, my bad, thinking it was downstream.

Comment (alyssawilk) [reply configs/google_com_http3_upstream_proxy.yaml:13]: oh I hoped the naming would be enough - let me add a comment at the top of the configs too!

**Review by yanavlasov — APPROVED** [review]

**Comment by triplewy** [comment]
@alyssawilk Is there an example h3 config for a quic listener that handles multiple certificates? Because quic listener does not support tls inspector filter so I'm not sure idiomatic way of routing multiple domains with one quic listener. For tcp listener, I would use tls inspector filter and filter chain matches on the sni to route requests to different http connection managers.

**Comment by danzh2010** [comment]
> @alyssawilk Is there an example h3 config for a quic listener that handles multiple certificates? Because quic listener does not support tls inspector filter so I'm not sure idiomatic way of routing multiple domains with one quic listener. For tcp listener, I would use tls inspector filter and filter chain matches on the sni to route requests to different http connection managers.

You can config multiple filter chains with different cert chains in transport socket. And you can config how filter chains are matched on connection properties.

**Comment by triplewy** [comment]
> 

@danzh2010 It seems for QUIC that you cannot do filter chain match on server_name because TLS inspector listener filter is not supported for QUIC. So it seems only option is to use one HttpConnectionManager for all QUIC connections and put all cert chains in one transport socket for QUIC

**Comment by triplewy** [comment]
@danzh2010 Ah so I tested with QUIC filter chain using server name and it seems that we don't need TLS inspector for server_name filter name match to work?
